README.md
# ring-hassio

## About
A fork of the Home Assistant add-on to enable live streams of Ring Cameras.
This add-on wraps around [Dgreif's excellent work](https://github.com/dgreif/ring) and exposes a livestream.

> **DOT NOT run this add-on with 24-hour streaming - use this addon by starting it on-demand**

## Installation
1. Add this GitHub repository to your **supervisor** (not HACS) add-on store. 
2. Configure your Ring Refresh Token and port (see configuration below).
3. Start the "Ring Livestream" add-on. Check for errors in the logs.
4. For remote access, open up the port in your router.
5. Open the stream at http://homeassistant.local:port/public/stream.m3u8 to make sure it works before going any further. We recommend using VLC or equivalent.
6. Add a camera to Home Assistant, such as:
   ```yaml
   camera:
     - platform: generic
       name: Ring Livestream
       stream_source: http://homeassistant.local:port/public/stream.m3u8
       still_image_url: http://homeassistant.local:port/public/stream.m3u8
    ```
    (Don't worry about the `still_image_url` not pointing to an actual image, we are not going to use it, but it is required.)
7. Add a card `Picture Glance` card to your UI, set the 'Camera Entity` to the camera you have just created.
8. Done! Enjoy your shiny new livestream!

## Configuration
Example configuration:
```yaml
ring_refresh_token: your_refresh_token
camera_name: Front Door
```
* You need to create a refresh token - see https://github.com/dgreif/ring/wiki/Refresh-Tokens on how to do that. Note that you will have to have node and npm installed on your machine.
* The camera name is the name entred when setting up the camera in the Ring app.

## Taking a snapshot
Currently the addon does not support taking snapshots, but when it does this is the configuration you will need:
In order to use the `snapshot` service, you will need to following settings in your `configuration.yaml`:
   ```yaml
   homeassistant:
     whitelist_external_dirs:
       - /config/tmp
   ```
   You can then call the `snapshot` service like this:
   ```yaml
   service: camera.snapshot
   entity_id: [entityID]
   filename: tmp/foo.jpg
   ```

Some users reported success to create a snapshot using:
```camera:
  - platform: ffmpeg
    input: http://hassio.local:port/public/stream.m3u8
```

## Battery conservation
A workaround to start/stop streaming (and avoid quick discharge) in Hassio is to start the addon on demand.
To do this set up a sensor and switch:

   ```yaml
   sensor:
     - platform: rest
       resource: "http://hassio.local:port/api/hassio/addons/xxxxxxxx_ringlivestream/info"
       headers:
         Authorization: "Bearer [Long_Lived_Access_Token]"
         Content-Type: application/json
       name: ring_addon_state
       value_template: "{{value_json['data']['state']}}" 
    
   switch:
     - platform: template
       switches:
         ring_live_stream:
           value_template: "{{ is_state('sensor.ring_addon_state', 'started') }}"
           turn_on:
             service: hassio.addon_start
             data:
               addon: xxxxxxxx_ringlivestream
           turn_off:
             service: hassio.addon_stop
             data:
               addon: xxxxxxxx_ringlivestream
   ```
- for the sensor you will need the URL of the Info page on the Ring Livestream add-on page.
- a `[Long_Lived_Access_Token]` can be generated on your profile page in Home Assistant (https://hassio.local:port/profile), scroll all the way to the bottom.
- To get the 8 chars code "xxxxxxxx_ringlivestream" get them from the same URL.

[patreon-shield]: https://frenck.dev/wp-content/uploads/2019/12/patreon.png
[patreon]: https://www.patreon.com/dutchdatadude


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
## [1.3.2] - 2023-09-19
Updated ring-client-api to 10.x. Updated Dockerfile for HA 2023. Updated 'run.sh'.

## [1.2] - 2020-07-13
Bumping package versions, ring-client-api to 9.6.0
 
## [1.1] - 2020-07-01
Bumping package versions

## [1.0] - 2020-06-10
We're bringing this addon back to life. It now supports 2FA authentication (Required by Ring). Please do not use this component for 24-hour streaming as it a) will consume your battery real quick and b) Ring will probably disconnect you (and all of us) from streaming pretty soon.

## [0.0.2] - 2020-02-06

### Changed
- Bumped Ring API component to latest version.

## [0.0.1] - 2020-02-05
### Added
- First release


ring_hassio/livestream.js
"use strict";
//Far majority of this code by Dgreif https://github.com/dgreif/ring/examples/browser_example.ts
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
exports.__esModule = true;
require("dotenv/config");
var ring_client_api_1 = require("ring-client-api");
var util_1 = require("util");
var fs = require('fs'), path = require('path'), http = require('http'), url = require('url'), zlib = require('zlib');
var PORT = process.env.RING_PORT;
//
var CAMERA_NAME = process.env.CAMERA_NAME;
var chosenCamera = CAMERA_NAME;
/**
 * This example creates an hls stream which is viewable in a browser
 * It also starts web app to view the stream at http://localhost:PORT
 **/
function startStream() {
    return __awaiter(this, void 0, void 0, function () {
        ///////////////
        function getCamera() {
            return __awaiter(this, void 0, void 0, function () {
                var cameras, camera, i, cameraName;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0: return [4 /*yield*/, ringApi.getCameras()];
                        case 1:
                            cameras = _a.sent();
                            //
                            if (chosenCamera) {
                                for (i = 0; i < cameras.length; i++) {
                                    cameraName = cameras[i].initialData.description;
                                    console.log("Checking If " + cameraName + " Is the same as the camera we are looking for (" + chosenCamera + ")");
                                    if (chosenCamera == cameraName) {
                                        camera = cameras[i];
                                        console.log("Matched " + cameraName);
                                    }
                                }
                            }
                            else {
                                camera = cameras[0];
                            }
                            //
                            if (!cameras) {
                                console.log('No cameras found');
                                return [2 /*return*/];
                            }
                            //
                            return [2 /*return*/, camera];
                    }
                });
            });
        }
        var ringApi, camera, publicOutputDirectory, server, sockets, nextSocketId, sipSession;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    ringApi = new ring_client_api_1.RingApi({
                        // Refresh token is used when 2fa is on
                        refreshToken: process.env.RING_REFRESH_TOKEN,
                        debug: true
                    });
                    return [4 /*yield*/, getCamera()
                        ///////////////
                    ];
                case 1:
                    camera = _a.sent();
                    publicOutputDirectory = path.join('public/');
                    console.log('output directory: ' + publicOutputDirectory);
                    server = http.createServer(function (req, res) {
                        // Get URL
                        var uri = url.parse(req.url).pathname;
                        console.log('requested uri: ' + uri);
                        // If Accessing The Main Page
                        if (uri == '/index.html' || uri == '/') {
                            res.writeHead(200, { 'Content-Type': 'text/html' });
                            res.write('<html><head><title>Ring Livestream</title></head><body>');
                            res.write('<h1>Welcome to your Ring Livestream!</h1>');
                            res.write("<video width=\"352\" height=\"198\" controls autoplay src=\"public/stream.m3u8\"></video>");
                            res.write("<br/>If you cannot see the video above open <a href=\"public/stream.m3u8\">the stream</a> in a player such as VLC.");
                            res.end();
                            return;
                        }
                        var filename = path.join("./", uri);
                        console.log('mapped filename: ' + filename);
                        fs.exists(filename, function (exists) {
                            if (!exists) {
                                console.log('file not found: ' + filename);
                                res.writeHead(404, { 'Content-Type': 'text/plain' });
                                res.write('file not found: %s\n', filename);
                                res.end();
                            }
                            else {
                                console.log('sending file: ' + filename);
                                switch (path.extname(uri)) {
                                    case '.m3u8':
                                        fs.readFile(filename, function (err, contents) {
                                            if (err) {
                                                res.writeHead(500);
                                                res.end();
                                            }
                                            else if (contents) {
                                                res.writeHead(200, { 'Content-Type': 'application/vnd.apple.mpegurl' });
                                                var ae = req.headers['accept-encoding'];
                                                if (ae && ae.match(/\bgzip\b/)) {
                                                    zlib.gzip(contents, function (err, zip) {
                                                        if (err)
                                                            throw err;
                                                        res.writeHead(200, { 'content-encoding': 'gzip' });
                                                        res.end(zip);
                                                    });
                                                }
                                                else {
                                                    res.end(contents, 'utf-8');
                                                }
                                            }
                                            else {
                                                console.log('empty playlist');
                                                res.writeHead(500);
                                                res.end();
                                            }
                                        });
                                        break;
                                    case '.ts':
                                        res.writeHead(200, { 'Content-Type': 'video/MP2T' });
                                        var stream = fs.createReadStream(filename, { bufferSize: 64 * 1024 });
                                        stream.pipe(res);
                                        break;
                                    default:
                                        console.log('unknown file type: ' +
                                            path.extname(uri));
                                        res.writeHead(500);
                                        res.end();
                                }
                            }
                        });
                    }).listen(PORT);
                    sockets = {}, nextSocketId = 0;
                    server.on('connection', function (socket) {
                        // Add a newly connected socket
                        var socketId = nextSocketId++;
                        sockets[socketId] = socket;
                        console.log('socket', socketId, 'opened');
                        // Remove the socket when it closes
                        socket.on('close', function () {
                            console.log('socket', socketId, 'closed');
                            delete sockets[socketId];
                        });
                        // Extend socket lifetime for demo purposes
                        socket.setTimeout(4000);
                    });
                    console.log('Started server, listening on port ' + PORT + '.');
                    return [4 /*yield*/, util_1.promisify(fs.exists)(publicOutputDirectory)];
                case 2:
                    if (!!(_a.sent())) return [3 /*break*/, 4];
                    return [4 /*yield*/, util_1.promisify(fs.mkdir)(publicOutputDirectory)];
                case 3:
                    _a.sent();
                    _a.label = 4;
                case 4: return [4 /*yield*/, camera.streamVideo({
                        output: [
                            '-preset',
                            'veryfast',
                            '-g',
                            '25',
                            '-sc_threshold',
                            '0',
                            '-f',
                            'hls',
                            '-hls_time',
                            '2',
                            '-hls_list_size',
                            '6',
                            '-hls_flags',
                            'delete_segments',
                            path.join(publicOutputDirectory, 'stream.m3u8')
                        ]
                    })];
                case 5:
                    sipSession = _a.sent();
                    sipSession.onCallEnded.subscribe(function () {
                        console.log('Call has ended');
                        server.close(function () { console.log('Server closed!'); });
                        // Destroy all open sockets
                        for (var socketId in sockets) {
                            console.log('socket', socketId, 'destroyed');
                            sockets[socketId].destroy();
                        }
                        //app.stop()
                        console.log('Restarting server');
                        startStream();
                    });
                    setTimeout(function () {
                        console.log('Stopping call...');
                        sipSession.stop();
                    }, 10 * 60 * 1000); // 10*60*1000 Stop after 10 minutes.
                    return [2 /*return*/];
            }
        });
    });
}
if (!('RING_REFRESH_TOKEN' in process.env) || !('RING_PORT' in process.env) || !('CAMERA_NAME' in process.env)) {
    console.log('Missing environment variables. Check RING_REFRESH_TOKEN, RING_PORT and CAMERA_NAME are set.');
    process.exit();
}
else {
    startStream();
}

ring_hassio/livestream.ts
//Far majority of this code by Dgreif https://github.com/dgreif/ring/examples/browser_example.ts

import 'dotenv/config'
import { RingApi } from 'ring-client-api'
import { promisify } from 'util'
const fs = require('fs'),
  path = require('path'),
  http = require('http'),
  url = require('url'),
  zlib = require('zlib')  

const PORT = process.env.RING_PORT;
//
const CAMERA_NAME = process.env.CAMERA_NAME;
var chosenCamera = CAMERA_NAME;
/**
 * This example creates an hls stream which is viewable in a browser
 * It also starts web app to view the stream at http://localhost:PORT
 **/

 async function startStream() {
   //
    const ringApi = new RingApi({
      // Refresh token is used when 2fa is on
      refreshToken: process.env.RING_REFRESH_TOKEN!,
      debug: true
    })
    ///////////////
    async function getCamera() {
      var cameras = await ringApi.getCameras();
      var camera;
      //
      if (chosenCamera) {
        for (var i=0; i < cameras.length; i++) {
          var cameraName = cameras[i].initialData.description;
          console.log(`Checking If ${cameraName} Is the same as the camera we are looking for (${chosenCamera})`);
          if (chosenCamera == cameraName) {
            camera = cameras[i];
            console.log(`Matched ${cameraName}`);
          } 
        }
      } else {
        camera = cameras[0]
      }
      //
      if (!cameras) {
        console.log('No cameras found')
        return
      }
      //
      return camera
  }
  ///////////////
  var camera = await getCamera()
  ///////////////
  const publicOutputDirectory = path.join('public/')
  console.log('output directory: '+publicOutputDirectory)

  var server = http.createServer(function (req, res) {
    // Get URL
    var uri = url.parse(req.url).pathname;
    console.log('requested uri: '+uri)
    // If Accessing The Main Page
    if (uri == '/index.html' || uri == '/') {
      res.writeHead(200, { 'Content-Type': 'text/html' });
      res.write('<html><head><title>Ring Livestream</title></head><body>');
      res.write('<h1>Welcome to your Ring Livestream!</h1>');
      res.write(`<video width="352" height="198" controls autoplay src="public/stream.m3u8"></video>`);
      res.write(`<br/>If you cannot see the video above open <a href="public/stream.m3u8">the stream</a> in a player such as VLC.`);
      res.end();
      return;
    }

    var filename = path.join("./", uri);
    console.log('mapped filename: '+filename)
	  fs.exists(filename, function (exists) {
		  if (!exists) {
		  	console.log('file not found: ' + filename);
		  	res.writeHead(404, { 'Content-Type': 'text/plain' });
		  	res.write('file not found: %s\n', filename);
		  	res.end();
		  }  else {
		   	  console.log('sending file: ' + filename);
		   	  switch (path.extname(uri)) {
		   	  case '.m3u8':
		  		  fs.readFile(filename, function (err, contents) {
					  if (err) {
					  	res.writeHead(500);
					  	res.end();
					  } else if (contents) {
					    res.writeHead(200,
					  	  {'Content-Type':
						    'application/vnd.apple.mpegurl'});
						  var ae = req.headers['accept-encoding'];
						  if (ae && ae.match(/\bgzip\b/)) {
						  	zlib.gzip(contents, function (err, zip) {
						  		if (err) throw err;
  								res.writeHead(200,
  								    {'content-encoding': 'gzip'});
  								res.end(zip);
  							});
  						} else {
  							res.end(contents, 'utf-8');
  						}
  					} else {
  						console.log('empty playlist');
  						res.writeHead(500);
  						res.end();
  					}
  				});
  				break;
  			case '.ts':
  				res.writeHead(200, { 'Content-Type':
  				    'video/MP2T' });
  				var stream = fs.createReadStream(filename,
  				    { bufferSize: 64 * 1024 });
  				stream.pipe(res);
          break;
	  		default:
	  			console.log('unknown file type: ' +
	  			    path.extname(uri));
	  			res.writeHead(500);
	  			res.end();
	  		}
	  	}
	});
  }).listen(PORT);

  // Maintain a hash of all connected sockets
  var sockets = {}, nextSocketId = 0;
  server.on('connection', function (socket) {
    // Add a newly connected socket
    var socketId = nextSocketId++;
    sockets[socketId] = socket;
    console.log('socket', socketId, 'opened');

    // Remove the socket when it closes
    socket.on('close', function () {
      console.log('socket', socketId, 'closed');
      delete sockets[socketId];
    });

    // Extend socket lifetime for demo purposes
    socket.setTimeout(4000);
  });
  console.log('Started server, listening on port '+PORT+'.')

  if (!(await promisify(fs.exists)(publicOutputDirectory))) {
    await promisify(fs.mkdir)(publicOutputDirectory)
  }
  //
  const sipSession = await camera.streamVideo({
    output: [
      '-preset',
      'veryfast',
      '-g',
      '25',
      '-sc_threshold',
      '0',
      '-f',
      'hls',
      '-hls_time',
      '2',
      '-hls_list_size',
      '6',
      '-hls_flags',
      'delete_segments',
      path.join(publicOutputDirectory, 'stream.m3u8')
    ]
  })

  sipSession.onCallEnded.subscribe(() => {
    console.log('Call has ended')
    server.close(function() {console.log('Server closed!');});
    // Destroy all open sockets
    for (var socketId in sockets) {
      console.log('socket', socketId, 'destroyed');
      sockets[socketId].destroy();
    }
    //app.stop()
    console.log('Restarting server')
    startStream()
  })

  setTimeout(function() {
    console.log('Stopping call...')
    sipSession.stop()
    
  }, 10* 60 * 1000) // 10*60*1000 Stop after 10 minutes.
}

if(!('RING_REFRESH_TOKEN' in process.env) || !('RING_PORT' in process.env) || !('CAMERA_NAME' in process.env)) {
  console.log('Missing environment variables. Check RING_REFRESH_TOKEN, RING_PORT and CAMERA_NAME are set.')
  process.exit()
}
else {
  startStream()
}




